rest/config_test.go

```diff
@@ -440,3 +443,370 @@ func TestAutoImportEnabled(t *testing.T) {
+
+	assert.Equal(t, *sc.DeprecatedLogFilePath, sc.Logging.LogFilePath)
+	assert.Equal(t, sc.DeprecatedLog, sc.Logging.Console.LogKeys)
+	assert.Len(t, warns, 2)
+}
+
+func TestSetupAndValidateLogging(t *testing.T) {
+	t.Skip("Skipping TestSetupAndValidateLogging")
+	defer base.SetUpTestLogging(base.LevelInfo, base.KeyAll)()
+	sc := &ServerConfig{}
+	warns, err := sc.SetupAndValidateLogging()
+	assert.NoError(t, err, "Setup and validate logging should be successful")
+	assert.NotEmpty(t, sc.Logging)
+	assert.Empty(t, sc.Logging.DeprecatedDefaultLog)
+	assert.Len(t, warns, 2)
+}
+
+func TestSetupAndValidateLoggingWithLoggingConfig(t *testing.T) {
+	t.Skip("Skipping TestSetupAndValidateLoggingWithLoggingConfig")
+	defer base.SetUpTestLogging(base.LevelInfo, base.KeyAll)()
+	logFilePath := "/var/log/sync_gateway"
+	logKeys := []string{"Admin", "Access", "Auth", "Bucket", "Cache"}
+	ddl := &base.LogAppenderConfig{LogFilePath: &logFilePath, LogKeys: logKeys, LogLevel: base.PanicLevel}
+	lc := &base.LoggingConfig{DeprecatedDefaultLog: ddl, RedactionLevel: base.RedactFull}
+	sc := &ServerConfig{Logging: lc}
+	warns, err := sc.SetupAndValidateLogging()
+	assert.NoError(t, err, "Setup and validate logging should be successful")
+	assert.Len(t, warns, 5)
+	assert.Equal(t, base.RedactFull, sc.Logging.RedactionLevel)
+	assert.Equal(t, ddl, sc.Logging.DeprecatedDefaultLog)
+}
+
+func TestServerConfigValidate(t *testing.T) {
+	defer base.SetUpTestLogging(base.LevelInfo, base.KeyAll)()
+	// unsupported.stats_log_freq_secs
+	statsLogFrequencySecs := uint(9)
+	unsupported := &UnsupportedServerConfig{StatsLogFrequencySecs: &statsLogFrequencySecs}
+	sc := &ServerConfig{Unsupported: unsupported}
+	assert.Len(t, sc.validate(), 1)
+
+	// Valid configuration value for StatsLogFrequencySecs
+	statsLogFrequencySecs = uint(10)
+	unsupported = &UnsupportedServerConfig{StatsLogFrequencySecs: &statsLogFrequencySecs}
+	sc = &ServerConfig{Unsupported: unsupported}
+	assert.Len(t, sc.validate(), 0)
+
+	// Explicitly disabled
+	statsLogFrequencySecs = uint(0)
+	unsupported = &UnsupportedServerConfig{StatsLogFrequencySecs: &statsLogFrequencySecs}
+	sc = &ServerConfig{Unsupported: unsupported}
+	assert.Len(t, sc.validate(), 0)
+}
+
+func TestSetupAndValidateDatabases(t *testing.T) {
+	// No error will be returned if the server config itself is nil
+	var sc *ServerConfig
+	errs := sc.setupAndValidateDatabases()
+	assert.Nil(t, errs)
+
+	// Simulate  invalid control character in URL while validating and setting up databases;
+	server := "walrus:\n\r"
+	bc := &BucketConfig{Server: &server}
+	databases := make(DbConfigMap, 2)
+	databases["db1"] = &DbConfig{Name: "db1", BucketConfig: *bc}
+
+	sc = &ServerConfig{Databases: databases}
+	errs = sc.setupAndValidateDatabases()
+	assert.Len(t, errs, 1)
+	assert.Contains(t, errs[0].Error(), "invalid control character in URL")
+}
+
+func TestParseCommandLine(t *testing.T) {
+	func() { config = nil }()
```

Comment:
As discussed, we should modify ParseCommandLine so that it returns a config that then can be used to update the global 'config', so that these tests don't have to worry about managing the global.

Reply:
This change has been implemented and committed to the latest revision.